Treat other items as functions for the purpose of type-based search

lolbinarycat opened the pull request:
specifically, constants and statics are nullary functions, and struct fields are unary functions.

fixes #130204

r? @notriddle

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11375071637/job/31645038742) [(plain)](https://github.com/rust-lang/rust/commit/0b4b586c2a1e755dd6f0928d2fb0c5479f02c9cf/checks/31645038742/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.......... (130/133)
..F        (133/133)


/checkout/tests/rustdoc-gui/search-tab.goml search-tab... FAILED
[ERROR] `tests/rustdoc-gui/search-tab.goml` line 82: The following errors happened: [` (7)  ` isn't equal to ` (6)  `]: for command `assert-text: ("#search-tabs > button:nth-child(2) > .count", " (6)  ")`
Error: ()
Build completed unsuccessfully in 0:03:40
  local time: Wed Oct 16 23:18:38 UTC 2024
  network time: Wed, 16 Oct 2024 23:18:38 GMT
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11375820999/job/31647213490) [(plain)](https://github.com/rust-lang/rust/commit/272fcc6fed1931484e5a04a3aaf2c326549ef9d0/checks/31647213490/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.99s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
Testing const-is-nullary-func.js ... OK
Testing deduplication.js ... OK
Testing enum-option.js ... OK
Testing exact-case.js ... OK
Testing field-is-unary-func.js ... tmp.js:4
    'others': [

SyntaxError: Unexpected string
SyntaxError: Unexpected string
    at internalCompileFunction (node:internal/vm:73:18)
    at wrapSafe (node:internal/modules/cjs/loader:1274:20)
    at Module._compile (node:internal/modules/cjs/loader:1320:27)
    at loadContent (/checkout/src/tools/rustdoc-js/tester.js:8:7)
    at runChecks (/checkout/src/tools/rustdoc-js/tester.js:376:24)
    at main (/checkout/src/tools/rustdoc-js/tester.js:545:29)
Command has failed. Rerun with -v to see more details.
  local time: Thu Oct 17 01:01:32 UTC 2024
  network time: Thu, 17 Oct 2024 01:01:32 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

lolbinarycat commented:
i just realized something: since this is implemented in rust code, unlike my other PRs, we don't have to make this insta-stable, it would actually be pretty easy to gate this behind a nightly feature (similarly to what is done with rustdoc-json)

i would honestly prefer running this as a t-rustdoc experiment first, since it still has a few unresolved questions (eg. how it will interact with generics)

personally, i would like to ship a MVP as a nightly-only feature, allowing me to start getting user feedback while i continue to iterate on it.

notriddle commented:
We could, technically, put this behind a `-Z` flag. But it wouldn't make much sense for docs.rs, since it's the reader who would want to choose, but it's the crate author who has the ability to set build parameters.

Making it a javascript Settings option would get it in front of more people.

lolbinarycat commented:
hmm.. i guess that would involve explicitly filtering only functions when doing type-based search unless that setting was enabled?

GuillaumeGomez commented:
That seems extremely strange to have constants and statics when we filter on functions.

lolbinarycat commented:
It's how hoogle works.  treating constants as pure nullary functions is not that unusual, especially if we're treating field accessors as unary functions.

when someone searches `-> SomeType`, what they're really asking is "how do i get a value of this type", to which a `new()` constructor is equally as valid of an answer as a constant.  it's not like they're actually writing a rust function signature, they don't have "fn" anywhere in their search.

the alternative is having different syntaxes for every way to get a value... which is more work both in implementation complexity, and also for users, who now have to perform 3 different queries to see every way they can construct a certain type.

GuillaumeGomez commented:
I see, thanks for the explanation. With this point of view, it indeed does make sense. The only requirement for this change for me will then to have documentation explaining what you just wrote. Sounds like a nice improvement.

lolbinarycat commented:
It seems like we're all in agreement that this should be done, but the next question is how?

insta-stable, settings panel, or nightly feature?

how i write the documentation will depend on how it is implemented.

GuillaumeGomez commented:
As far as I know, any change in search is insta-stable.

As for how to implement it, what are the different ways you see?

lolbinarycat commented:
> As far as I know, any change in search is insta-stable.

well, since this is actually implemented in rust, and not js, we do actually have easy access to feature gates, although it might be a bit confusing.

> As for how to implement it, what are the different ways you see?

I just mean should it be gated behind anything, because if it is, i'll add an extra section for it, but if not, i'll replace/modify the existing type-based search section.

notriddle commented:
Changes to search are usually insta-stable. It's not an API, so if something turns out to be a bad idea, we can undo it.

lolbinarycat commented:
somehow #132123 didn't cause a merge conflict??

camelid commented:
I can see this making sense for consts and statics, but it feels like a stretch to consider struct fields as unary functions -- it also doesn't seem useful.

lolbinarycat commented:
> I can see this making sense for consts and statics, but it feels like a stretch to consider struct fields as unary functions -- it also doesn't seem useful.

the point is to improve devx in the case where you are looking for a getter method, but there isn't one, since you are expected to just access the field directly.

this is especially common with things like `rustdoc::Context`, where direct field access and getter methods are used in equal measure.

as i explained to GG:
1. this improves parity with hoogle, which treats struct fields as unary functions.   people who are already familiar with hoogle are those most likely to use the advanced features of rustdoc search, so we should try to avoid friction between the behavior of the two tools.
2. when people search `Foo -> Bar`, what they're really asking is "how do I get `Bar` out of `Foo`?".  usually this will be a getter method, but sometimes it will be a field. (sometimes it will require a chain of multiple operations, but at least in that case, `-> Bar` will now show any fields that might be a relevant pathway)

GuillaumeGomez commented:
I agreed for consts and statics, not for fields. ^^' I agree with @camelid: it's too much of a fetch here.

I'm not sure trying to reach parity with hoogle should be a goal for rustdoc either.

notriddle commented:
> I agreed for consts and statics, not for fields.

But why consts and statics, then? What makes them different from fields?

camelid commented:
Probably the biggest concern I have with fields is it feels like noise. For most structs, you already know what the fields are just from the sidebar of the struct's page. When I search for e.g. `rustdoc::Context ->` I want to see what functions do stuff with it, not all its fields -- since those are easy to find.

I don't feel super strongly about this, but comparing it to Hoogle is a little bit of a false comparison because in Haskell fields are _actually_ functions (and same with constants). In Rust, it's more of an analogy -- we don't have `.field(my_struct)` syntax, for example.

lolbinarycat commented:
> When I search for e.g. `rustdoc::Context ->` I want to see what functions do stuff with it, not all its fields -- since those are easy to find.

Hm, I hadn't considered this form of search.  I would say we can de-prioritize fields when no return type is given, but I'd like to wait and do that in a follow-up PR.

notriddle commented:
It would certainly be possible to exclude fields from the single-element type-based queries:

```diff
diff --git a/src/librustdoc/html/static/js/search.js b/src/librustdoc/html/static/js/search.js
index eed64d024c0..f41975bf759 100644
--- a/src/librustdoc/html/static/js/search.js
+++ b/src/librustdoc/html/static/js/search.js
@@ -52,6 +52,7 @@ const itemTypes = [
 // used for special search precedence
 const TY_GENERIC = itemTypes.indexOf("generic");
 const TY_IMPORT = itemTypes.indexOf("import");
+const TY_STRUCTFIELD = itemTypes.indexOf("structfield");
 const ROOT_PATH = typeof window !== "undefined" ? window.rootPath : "../";
 
 // Hard limit on how deep to recurse into generics when doing type-driven search.
@@ -2984,7 +2985,7 @@ class DocSearch {
                 fullId,
                 parsedQuery.typeFingerprint,
             );
-            if (tfpDist !== null) {
+            if (tfpDist !== null && row.ty !== TY_STRUCTFIELD) {
                 const in_args = row.type && row.type.inputs
                     && checkIfInList(row.type.inputs, elem, row.type.where_clause, null, 0);
                 const returned = row.type && row.type.output
@@ -3285,6 +3286,9 @@ class DocSearch {
                 parsedQuery.elems.sort(sortQ);
                 parsedQuery.returned.sort(sortQ);
                 for (let i = 0, nSearchIndex = this.searchIndex.length; i < nSearchIndex; ++i) {
+                    if (this.searchIndex[i].ty === TY_STRUCTFIELD && parsedQuery.foundElems === 1) {
+                        continue;
+                    }
                     handleArgs(this.searchIndex[i], i, results_others);
                 }
             }
```

The other option is to make the ranking algorithm sort them lower.

```diff
diff --git a/src/librustdoc/html/static/js/search.js b/src/librustdoc/html/static/js/search.js
index eed64d024c0..349db638986 100644
--- a/src/librustdoc/html/static/js/search.js
+++ b/src/librustdoc/html/static/js/search.js
@@ -52,6 +52,7 @@ const itemTypes = [
 // used for special search precedence
 const TY_GENERIC = itemTypes.indexOf("generic");
 const TY_IMPORT = itemTypes.indexOf("import");
+const TY_STRUCTFIELD = itemTypes.indexOf("structfield");
 const ROOT_PATH = typeof window !== "undefined" ? window.rootPath : "../";
 
 // Hard limit on how deep to recurse into generics when doing type-driven search.
@@ -2144,6 +2144,13 @@ class DocSearch {
                     return a - b;
                 }
 
+                // sort fields lower in type-based search
+                a = aaa.item.ty === TY_STRUCTFIELD;
+                b = bbb.item.ty === TY_STRUCTFIELD;
+                if (a !== b && isType) {
+                    return a - b;
+                }
+
                 // Sort by distance in the name part, the last part of the path
                 // (less changes required to match means higher rankings)
                 a = (aaa.dist);
```

lolbinarycat commented:
> It would certainly be possible to exclude fields from the single-element type-based queries:

your code snippet seems like it would also exclude `-> SomeType` queries, which i think would be a very bad idea.

> All else being equal, I prefer ranking.

Indeed, and I believe ranking is something we should iterate on in several other aspects.

bors commented:
:umbrella: The latest upstream changes (presumably #134414) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez commented:
We discussed about it the last [rustdoc team meeting](https://rust-lang.zulipchat.com/#narrow/channel/393423-t-rustdoc.2Fmeetings/topic/2025-01-13) and we agreed with this change.

However, the items should be ranked lower than functions and methods.

Quoting @camelid:

> in my experience, looking for something that ends up being a field has never taken me very long -- but that's just my personal experience. of course it's different for different people in different contexts

lolbinarycat committed:
Treat other items as functions for the purpose of type-based search

constants and statics are nullary functions, and struct fields are unary functions.

functions (along with methods and trait methods) are prioritized over other
items, like fields and constants.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12756518236/job/35554826397) [(plain)](https://github.com/rust-lang/rust/commit/cbc94ef70512f7acdff0c9d062a4cfea7246458e/checks/35554826397/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rinja v0.3.5
error[E0308]: `match` arms have incompatible types
   --> src/librustdoc/html/render/search_index.rs:843:39
    |
836 |       let (mut inputs, mut output, param_names, where_clause) = match item.kind {
    |  _______________________________________________________________-
837 | |         clean::ForeignFunctionItem(ref f, _)
838 | |         | clean::FunctionItem(ref f)
839 | |         | clean::MethodItem(ref f, _)
840 | |         | clean::RequiredMethodItem(ref f) => {
841 | |             get_fn_inputs_and_outputs(f, tcx, impl_or_trait_generics, cache)
    | |             ---------------------------------------------------------------- this is found to be of type `(Vec<RenderType>, Vec<RenderType>, Vec<rustc_span::Symbol>, Vec<Vec<RenderType>>)`
842 | |         }
843 | |         clean::ConstantItem(ref c) => make_nullary_fn(&c.type_),
    | |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^ expected a tuple with 4 elements, found one with 3 elements
859 | |         _ => return None,
860 | |     };
    | |_____- `match` arms have incompatible types
    |
    |
    = note: expected tuple `(Vec<RenderType>, Vec<RenderType>, Vec<rustc_span::Symbol>, Vec<Vec<RenderType>>)`
               found tuple `(Vec<RenderType>, Vec<RenderType>, Vec<Vec<RenderType>>)`
For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustdoc` (lib) due to 1 previous error
Build completed unsuccessfully in 0:02:31
  local time: Mon Jan 13 21:48:10 UTC 2025
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12757450226/job/35557735808) [(plain)](https://github.com/rust-lang/rust/commit/dada2e1cc70ba73dfab1aaa012198f2567cc4f86/checks/35557735808/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.972 Building wheels for collected packages: reuse
#13 2.973   Building wheel for reuse (pyproject.toml): started
#13 3.185   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.186   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=be6760d5849de4a58bbe52b85ca57a55f2b32b518b17029a5ad2e530db0d4303
#13 3.186   Stored in directory: /tmp/pip-ephem-wheel-cache-8mn5dcjt/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.189 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.596 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.597 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.149 Collecting virtualenv
#13 4.149 Collecting virtualenv
#13 4.200   Downloading virtualenv-20.28.1-py3-none-any.whl (4.3 MB)
#13 4.360      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 27.1 MB/s eta 0:00:00
#13 4.424 Collecting filelock<4,>=3.12.2
#13 4.432   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.456 Collecting distlib<1,>=0.3.7
#13 4.463   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.476      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 44.5 MB/s eta 0:00:00
#13 4.513 Collecting platformdirs<5,>=3.9.1
#13 4.521   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.603 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.790 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.28.1
#13 DONE 4.9s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      131008 kB
DirectMap2M:     7208960 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-01-08/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
fmt: checked 5772 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/src/librustdoc/html/static/js/search.js:2753: tab character
##[error]tidy error: /checkout/src/librustdoc/html/static/js/search.js:2754: tab character
##[error]tidy error: /checkout/src/librustdoc/html/static/js/search.js:2755: tab character
##[error]tidy error: /checkout/src/librustdoc/html/static/js/search.js:2756: tab character
##[error]tidy error: /checkout/src/librustdoc/html/static/js/search.js:2757: tab character
##[error]tidy error: /checkout/src/librustdoc/html/static/js/search.js:2758: tab character
##[error]tidy error: /checkout/src/librustdoc/html/static/js/search.js:2759: tab character
##[error]tidy error: /checkout/src/librustdoc/html/static/js/search.js:2760: tab character
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking python file formatting
28 files already formatted
checking C++ file formatting
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Mon Jan 13 22:49:47 UTC 2025
  network time: Mon, 13 Jan 2025 22:49:47 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12770187997/job/35594460913) [(plain)](https://github.com/rust-lang/rust/commit/fb2c74fe45e3274c35f0b465384762493296c7d4/checks/35594460913/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rinja v0.3.5
error[E0308]: `match` arms have incompatible types
   --> src/librustdoc/html/render/search_index.rs:843:39
    |
836 |       let (mut inputs, mut output, param_names, where_clause) = match item.kind {
    |  _______________________________________________________________-
837 | |         clean::ForeignFunctionItem(ref f, _)
838 | |         | clean::FunctionItem(ref f)
839 | |         | clean::MethodItem(ref f, _)
840 | |         | clean::RequiredMethodItem(ref f) => {
841 | |             get_fn_inputs_and_outputs(f, tcx, impl_or_trait_generics, cache)
    | |             ---------------------------------------------------------------- this is found to be of type `(Vec<RenderType>, Vec<RenderType>, Vec<rustc_span::Symbol>, Vec<Vec<RenderType>>)`
842 | |         }
843 | |         clean::ConstantItem(ref c) => make_nullary_fn(&c.type_),
    | |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^ expected a tuple with 4 elements, found one with 3 elements
859 | |         _ => return None,
860 | |     };
    | |_____- `match` arms have incompatible types
    |
    |
    = note: expected tuple `(Vec<RenderType>, Vec<RenderType>, Vec<rustc_span::Symbol>, Vec<Vec<RenderType>>)`
               found tuple `(Vec<RenderType>, Vec<RenderType>, Vec<Vec<RenderType>>)`
For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustdoc` (lib) due to 1 previous error
Build completed unsuccessfully in 0:02:32
  local time: Tue Jan 14 15:02:04 UTC 2025
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12776588068/job/35615576948) [(plain)](https://github.com/rust-lang/rust/commit/7af6e4c23100699e10e7f5db9d753238edab51d0/checks/35615576948/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: ES-Check: there were no ES version matching errors!  🎉
+ eslint -c ../src/librustdoc/html/static/.eslintrc.js ../src/librustdoc/html/static/js/externs.js ../src/librustdoc/html/static/js/main.js ../src/librustdoc/html/static/js/scrape-examples.js ../src/librustdoc/html/static/js/search.js ../src/librustdoc/html/static/js/settings.js ../src/librustdoc/html/static/js/src-script.js ../src/librustdoc/html/static/js/storage.js

/checkout/src/librustdoc/html/static/js/search.js
  2755:65  error  'TY_METHOD' is not defined  no-undef
  2756:65  error  'TY_METHOD' is not defined  no-undef

✖ 2 problems (2 errors, 0 warnings)
  local time: Tue Jan 14 21:53:44 UTC 2025
  network time: Tue, 14 Jan 2025 21:53:44 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12777268911/job/35617740162) [(plain)](https://github.com/rust-lang/rust/commit/258d4dd9e00fca08afacefadd7301e64093878b0/checks/35617740162/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#22 exporting to docker image format
#22 sending tarball 27.8s done
#22 DONE 34.2s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.97s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
Testing alias-4.js ... OK
Testing alias.js ... OK
Testing asrawfd.js ... OK
Testing basic.js ... OK
Testing bufread-fill-buf.js ... TypeError: Cannot read properties of undefined (reading 'ty')
    at /checkout/obj/build/x86_64-unknown-linux-gnu/doc/static.files/search-13b7acc0.js:1:43399
    at Array.sort (<anonymous>)
    at sortResults (/checkout/obj/build/x86_64-unknown-linux-gnu/doc/static.files/search-13b7acc0.js:1:43070)
    at DocSearch.execQuery (/checkout/obj/build/x86_64-unknown-linux-gnu/doc/static.files/search-13b7acc0.js:1:65468)
    at Object.doSearch (/checkout/src/tools/rustdoc-js/tester.js:440:44)
    at doSearch (/checkout/src/tools/rustdoc-js/tester.js:559:31)
    at runSearch (/checkout/src/tools/rustdoc-js/tester.js:219:27)
    at /checkout/src/tools/rustdoc-js/tester.js:380:20
    at runCheckInner (/checkout/src/tools/rustdoc-js/tester.js:308:28)
    at runCheck (/checkout/src/tools/rustdoc-js/tester.js:335:24)
Command has failed. Rerun with -v to see more details.
  local time: Tue Jan 14 22:49:00 UTC 2025
  network time: Tue, 14 Jan 2025 22:49:00 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

notriddle commented on src/librustdoc/html/static/js/search.js:
```diff
@@ -2749,6 +2751,15 @@ class DocSearch {
 
+                // in type based search, put functions first
+                if (parsedQuery.hasReturnArrow) {
+                    a = (aaa.item.ty !== TY_FN && a.item.ty !== TY_METHOD);
+                    b = (bbb.item.ty !== TY_FN && b.item.ty !== TY_METHOD);
```
Sorry. I got this wrong...

```suggestion
                    a = (aaa.item.ty !== TY_FN && aaa.item.ty !== TY_METHOD);
                    b = (bbb.item.ty !== TY_FN && bbb.item.ty !== TY_METHOD);
```

notriddle commented:
@bors r+ rollup

bors commented:
:pushpin: Commit a46b7e10b5163342289d7d2a33d4c4febd15e584 has been approved by `notriddle`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle a46b7e10b5163342289d7d2a33d4c4febd15e584 -->
<!-- homu: {"type":"Approved","sha":"a46b7e10b5163342289d7d2a33d4c4febd15e584","approver":"notriddle","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

notriddle commented:
Wait a minute.

@bors r-

Can you please squash these commits? Then r=me

lolbinarycat commented:
r=@notriddle

GuillaumeGomez commented:
@bors r=notriddle rollup

bors commented:
:pushpin: Commit fe6deb0fa146b762ba809354371518793578e744 has been approved by `notriddle`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle fe6deb0fa146b762ba809354371518793578e744 -->
<!-- homu: {"type":"Approved","sha":"fe6deb0fa146b762ba809354371518793578e744","approver":"notriddle","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

GuillaumeGomez commented:
I think it's the one which failed in https://github.com/rust-lang/rust/pull/135533.

@bors r-

lolbinarycat commented:
i'm pretty sure this is a bug in `tester.js`, specifically this line:

```js
                error_text.push(queryName + "==> '" + JSON.stringify(elem) + "' was supposed " +
                                "to be before '" + JSON.stringify(results[key][entry_pos]) + "'");
```

the second part should be using `prev_pos` instead of `entry_pos`.  currently it prints:
1. the current entry, as it appears in `EXPECTED`.
2. the current entry, as it appears in the search results.

this conflicts with what the error message suggests is printed.

notriddle commented:
@bors r+

bors commented:
:pushpin: Commit 9397d133f683070863029552313b5e083398bf42 has been approved by `notriddle`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle 9397d133f683070863029552313b5e083398bf42 -->
<!-- homu: {"type":"Approved","sha":"9397d133f683070863029552313b5e083398bf42","approver":"notriddle","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
